Step1 - 문자열계산기

junseong94 opened the pull request:
안녕하세요 리뷰어님. 수강생 김준성 입니다.

1단계 문자열 덧셈 계산기 개발이 완료되어 Pull Request 합니다.

과제 진행 중 질문 사항과 피드백을 받고자 하는 부분을 아래와 같이 전달드립니다.

1. 음수 입력 조건을 처리하고자 정규식을 사용하였는데 해결 가능한 다른 방식이 있을까요?
2. add 함수에 너무 많은 역할이 주어진 것 같은데 해당 코드에 대해 자세히 피드백 받고 싶습니다.

junseong94 committed:
add StringCalculatorTest

junseong94 committed:
refactor(sum) : use split()

junseong94 committed:
add custom delimiter

junseong94 committed:
rafactor: all methods
add test for negative number

liquidjoo commented on src/test/java/stringcalculator/StringCalculator.java:
```diff
@@ -0,0 +1,44 @@
+    public StringCalculator() {
+    }
+
+    public int add(String text) {
```
질문 주신 내용 중 많은 역할, 책임이 존재하는 것 같아보이네요~
책임을 메서드 단위로 나누어 보고
비슷한 책임을 묶어 클래스를 만들어 역할을 나누어 보면 어떨까요?

liquidjoo replied:
add 에 아직 많은 책임이 있는 것 같아요 :)
생성자를 활용해 초기화 시 유효성 검사를 해보고
다른 객체를 만들어 그 객체에 맞는 책임을 부여하여 나누어 보면 어떨까요?

liquidjoo commented on src/test/java/stringcalculator/StringCalculator.java:
```diff
@@ -0,0 +1,44 @@
+            return 0;
+        }
+
+        // 음수 입력 시 예외 처리
+        if(text.matches("^-\\d+$")){
+            throw new RuntimeException();
+        }
```
매개 변수에 대한 유효성 검증이네요 :)
상태값으로 분리해보고 이를 검증하는 것을 만들어 보면 좋을 것 같아요
지금도 충분히 실행되는 코드이지만 유연하게 개발하고 싶다면 확장성에 대해서 생각해보면 더욱 좋습니다 :)

liquidjoo commented on src/test/java/stringcalculator/StringCalculator.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    public Matcher delimiter(String text) {
+        return Pattern.compile("//(.)\n(.*)").matcher(text);
```
패턴은 메모리를 많이 잡아 먹게 되는 유틸성 객체인데요 :)
이를 어떻게 효율적으로 만들 수 있을까요?

liquidjoo commented on src/test/java/stringcalculator/StringCalculatorTest.java:
```diff
@@ -0,0 +1,74 @@
+
+public class StringCalculatorTest {
+
+    /**
+     * 요구 사항
+     *
+     * 쉼표(,) 또는 콜론(:)을 구분자로 가지는 문자열을 전달하는 경우 구분자를 기준으로 분리한 각 숫자의 합을 반환 (예: “” => 0, "1,2" => 3, "1,2,3" => 6, “1,2:3” => 6)
+     * 앞의 기본 구분자(쉼표, 콜론) 외에 커스텀 구분자를 지정할 수 있다. 커스텀 구분자는 문자열 앞부분의 “//”와 “\n” 사이에 위치하는 문자를 커스텀 구분자로 사용한다.
+     * 예를 들어 “//;\n1;2;3”과 같이 값을 입력할 경우 커스텀 구분자는 세미콜론(;)이며, 결과 값은 6이 반환되어야 한다.
+     *
+     * 문자열 계산기에 숫자 이외의 값 또는 음수를 전달하는 경우 RuntimeException 예외를 throw 한다.
+     */
```
요구 사항들은 테스트의 `DisplayName` 을 통해 충분히 나타낼 수 있기에 주석은 없어도 될 것 같아요~

liquidjoo commented on src/test/java/stringcalculator/StringCalculator.java:
```diff
@@ -0,0 +1,44 @@
+
+        Matcher matcher = delimiter(text);
+
+        if (matcher.find()) {
+            String delimiter = matcher.group(1);
+            split = matcher.group(2).split(delimiter);
+        }else{
+            split = text.split(",|:");
+        }
```
else 를 사용하지 않고 구현해보면 좋을 것 같아요~

liquidjoo reviewed (changes requested):
안녕하세요 준성님 :)
1단계 미션 구현 잘 해주셨어요 👍 
질문 주신 내용들 참고하여 코멘트 남겼습니다~
확인하고 다시 요청 주세요 🙇 
궁금한 사항은 언제든 DM 보내주세요~

junseong94 committed:
rafactor(add method) : Reorganized responsibility into new classes for role separation
rafactor(negative regex) : used by Enum
rafactor(Pattern Object) : used by static final
refactor(comments) : deleted unnessary comments
refactor(else statement) : Initialize previously else code upfront

liquidjoo commented on src/test/java/stringcalculator/DelimiterParser.java:
```diff
@@ -0,0 +1,28 @@
+
+public class DelimiterParser {
+
+    // Pattern을 static final로 선언해 클래스 로드 시 한번만 컴파일되도록 개선
```
주석으로 남기는 것도 좋지만 커밋을 통해 남기는 방향이 더 올바른 것 같습니다 :)
다음엔 주석을 제거해주세요~

liquidjoo commented on src/test/java/stringcalculator/StringCalculator.java:
```diff
@@ -8,37 +8,33 @@ public class StringCalculator {
 
     private String text;
```
사용하고 있지 않는 것 같은데요~
생성자를 활용하면 어떻게 활용할 수 있을까요?

liquidjoo commented on src/test/java/stringcalculator/StringValidation.java:
```diff
@@ -0,0 +1,35 @@
+package stringcalculator;
+
+enum ValidationNumber {
+
+    NEGATIVE_NUM_REGEX("^-\\d+$"),
+    POSITIVE_NUM_REGEX("^\\d+$");
+
+    private final String regex;
+
+    ValidationNumber(String regex) {
+        this.regex = regex;
+    }
+    public String getRegex() {
+        return regex;
+    }
+
+}
```
enum 을 사용한다면 별도의 클래스로 분리해주면 좋을 것 같아요 :)

liquidjoo commented on src/test/java/stringcalculator/StringCalculator.java:
```diff
@@ -8,37 +8,33 @@ public class StringCalculator {
     private String text;
 
-    public StringCalculator() {
+    private static StringValidation stringValidation = new StringValidation();
+    private static DelimiterParser delimiterParser = new DelimiterParser();
```
객체지향설계, 개발에서 static 사용을 지양하고 있는데요
어떤 이유인지 알아보면 좋을 것 같아요 :)

liquidjoo commented on src/test/java/stringcalculator/StringValidation.java:
```diff
@@ -0,0 +1,35 @@
+
+}
+
+public class StringValidation {
```
생성자를 활용하면 어떨까요?

liquidjoo commented on src/test/java/stringcalculator/StringValidation.java:
```diff
@@ -0,0 +1,35 @@
+
+    public void checkNegative(String[] numbers) {
+        for (String number : numbers) {
+            if(number.matches(negativeValid)) {
+                throw new RuntimeException();
+            }
```
메서드를 분리하면 뎁스도 줄여볼 수 있고 테스트 코드 또한 작성이 편할 것 같아요 :)

liquidjoo replied:
새롭게 만들어진 책임, 역할에 대해서 테스트 코드 작성해보면 좋을 것 같아요~

liquidjoo reviewed (changes requested):
안녕하세요 준성님 :)
1단게 피드백 잘해주셨어요~
몇 가지 코멘트 남겼습니다 확인하고 다시 요청 주세요 🙇

junseong94 committed:
 ✔  at 19:50:47 
dquote> rafactor(StringCalculator) : updated constructor(incl. StringValidation), removed related code of StringValidation
dquote> rafactor(DelimiterParser) : updated Pattern initialized code
dquote> Add(DelimiterParserTest) : Add the test code for DelimiterParser
dquote> rafactor(StringValidation) : separated enum to ValidationRegex class, update constructor(incl. DelimiterParser parameter), add parseNumber method
dquote> Add(StringValidationTest) : Add the test code for DelimiterParser

junseong94 committed:
delete(comments) : unnessary comments

junseong94 commented:
안녕하세요 리뷰어님
질문 주신 것 중 static지양하는 이유에 대해 아래와 같이 정리해 보았습니다.
덕분에 처음으로 static에 대해 알아보면서 몰랐던 사항들이 많아 다시한번 공부하게된 기회기 되었습니다. 
감사합니다. 리뷰어님 :) 

* 객체지향설계, 개발에서 static 지양하는 이유 : 
- static 은 전역 상태에 영향을 미치기에 전체 프로그램에서 호출할 수 있습니다. 이는 코드의 여러 부분에 영향을 미칠 수 있기에 이슈 발생 시 추적이 힘듭니다.
- static으로 선언한 메서드와 필드는 static 영역에 클래스로더 시점에 메모리에 올라가 Garbage Collector의 메모리 관리를 받지 않기에 메모리 회수를 할 수 없어 메모리 누수가 발생할 수 있습니다.

* static 대안 :
1. “transient”나 “volatile”과 같은 제어자(modifier)를 사용
2. 실행 속도를 빠르게 해주는 메소드 인라이닝을 위해 “final” 메서드를 사용

liquidjoo commented on src/test/java/stringcalculator/StringValidation.java:
```diff
@@ -1,35 +1,36 @@
+
+        String[] numbers = delimiterParser.parseDelimiter(text);
+
+        return Arrays.stream(numbers).mapToInt(Integer::parseInt).sum();
```
유효성 검증에서 Number 를 파싱하는 것은 조금 어색해 보여요~
또한 검증을 위한 유효성 로직이 있어 단일책임원칙에 어긋나 보이네요

liquidjoo commented on src/test/java/stringcalculator/StringCalculator.java:
```diff
@@ -1,40 +1,15 @@
 
     public StringCalculator(){
+        this.stringValidation = new StringValidation();
     }
```
```suggestion
    public StringCalculator(String text){
        this.stringValidation = new StringValidation(text);
        this.text = new Parser(text).getText;
        ...
    }
```

단편적인 예를 표현한 것이지만 상태 값을 활용해 보면 어떨까요?

liquidjoo commented on src/test/java/stringcalculator/StringValidation.java:
```diff
@@ -1,35 +1,36 @@
+public class StringValidation {
 
-    private final String regex;
+    private final DelimiterParser delimiterParser;
```
유효성 검증에 Parser 는 어울리지 않는 것 같아요~
오히려 객체의 결합도가 올라간 것 같네요
어떻게 해결해 볼 수 있을까요?

liquidjoo reviewed (changes requested):
안녕하세요 준성님 :)
1단계 피드백 잘해주셨어요 👍 
몇 가지 코멘트 남겼습니다~ 확인하고 다시 요청해 주세요 🙇 
궁금한 사항은 언제든 DM 보내주세요

junseong94 committed:
refactor(test class) : StringCalculationTest, StringValidationTest
extract(delimiter regex) : Extract regex pattern of custom delimiter into ValidationRegex Enum
rafactor(constructor) : add init codes of StringValidation & DelimiterParser in a constructor of StringCalculator
delete(init code) : delete init code of DelimiterParser at StringValidation

liquidjoo commented on src/test/java/stringcalculator/DelimiterParser.java:
```diff
@@ -5,7 +5,10 @@
 
-    private final Pattern regexPattern = Pattern.compile("//(.)\n(.*)");
+    private final Pattern regexPattern = Pattern.compile(ValidationRegex.CUSTOM_DELIMITER_REGEX.getRegex());
+
+    public DelimiterParser() {
+    }
 
     public String[] parseDelimiter(String text) {
```
```suggestion
    private String[] parsedNumbers;
    public DelimiterParser(String text) {
        this.parsedNumbers = parseDelimiter(text);
    }
   
    private String[] parseDelimiter(String text) {
```

와 같이 상태값과 초기화 할 수 있는 것을 활용해보면 어떨까요?

liquidjoo commented on src/test/java/stringcalculator/StringValidation.java:
```diff
@@ -1,13 +1,14 @@
 
-    public StringValidation() {
-        this.delimiterParser = new DelimiterParser();
+    public StringValidation(String text) {
+        this.text = text;
```
```suggestion
    public StringValidation(String text) {
        validateNum(text);
        this.text = text;
```

와 같이 표현해 볼 수 있지 않을까요?

liquidjoo reviewed (changes requested):
안녕하세요 준성님 :)
1단계 피드백 잘 반영해주셨어요
몇 가지 코멘트 남겼어요~ 확인하고 다시 요청주세요 🙇 
궁금한 사항이 있으면 언제든 DM 보내주세요~

junseong94 committed:
refactor(constructor) : StringValidation, DelimiterParser
update(Test code) : Updated test code due to constructor refactoring

liquidjoo commented on src/test/java/stringcalculator/DelimiterParser.java:
```diff
@@ -6,8 +6,10 @@
+    private String[] parsedNumbers;
 
-    public DelimiterParser() {
+    public DelimiterParser(String text) {
+        this.parsedNumbers = parseDelimiter(text);
```
피드백 반영 좋습니다 👍

liquidjoo commented on src/test/java/stringcalculator/StringCalculator.java:
```diff
@@ -8,8 +8,8 @@ public class StringCalculator {
+        this.parsedNum = new DelimiterParser(parseReulst).getParsedNumber();
     }
 
     public int add() {
```
해당 객체의 아쉬움이 하나있는데요! 
add 와 결과를 보여주는 역할을 나누어서 만들어보면 어떨까요~?

liquidjoo reviewed (changes requested):
안녕하세요 준성님 :)
1단계 피드백 잘해주셨어요 👍 
한 가지 코멘트 남겼어요~ 확인하고 다시 요청주세요~
다음 요청시엔 머지를 진행하도록 하겠습니다 🙇

junseong94 committed:
refactor(methods) : add showResult Method for showing calculated results, updated return type of add() from int to void

liquidjoo commented on src/test/java/stringcalculator/StringCalculator.java:
```diff
@@ -5,17 +5,19 @@
-       return Arrays.stream(parsedNum).mapToInt(Integer::parseInt).sum();
+    public void add() {
+        addResult = Arrays.stream(parsedNum).mapToInt(Integer::parseInt).sum();
     }
 
-
-
+    public int showResult() {
+        return addResult;
+    }
```
피드백 반영 잘해주셨어요 👍

liquidjoo reviewed (approved):
안녕하세요 준성님 :)
1단계 피드백 반영 정말 잘해주셨어요 👍 
다음 단계 진행해 주세요 🙇
